aeppl/truncation.py

```diff
@@ -122,3 +137,216 @@ def censor_logprob(op, values, base_rv, lower_bound, upper_bound, **kwargs):
+
+    if rng is None:
+        rng = shared(np.random.RandomState(), borrow=True)
+
+    # Try to use specialized Op
+    try:
+        truncated_rv, updates = _truncated(
+            rv.owner.op, lower, upper, rng, *rv.owner.inputs[1:]
+        )
+        return truncated_rv, updates
+    except NotImplementedError:
+        pass
+
+    # Variables with `_` suffix identify dummy inputs for the OpFromGraph
+    # We will use the Shared RNG variable directly because Scan demands it, even
+    # though it would not be necessary for the icdf OgFromGraph
+    graph_inputs = [*rv.owner.inputs[1:], lower, upper]
+    graph_inputs_ = [inp.type() for inp in graph_inputs]
+    *rv_inputs_, lower_, upper_ = graph_inputs_
+
+    # Try to use inverted cdf sampling
+    try:
+        rv_ = rv.owner.op.make_node(rng, *rv_inputs_).default_output()
+        rv_.name = rv.name
+        # For left truncated discrete RVs, we need to include the whole lower bound.
+        # This may result in draws below the truncation range, if any uniform == 0
+        lower_value = lower_ - 1 if rv.owner.op.dtype.startswith("int") else lower_
+        cdf_lower_ = at.exp(logcdf(rv_, lower_value))
+        cdf_upper_ = at.exp(logcdf(rv_, upper_))
+        uniform_ = at.random.uniform(
+            cdf_lower_,
+            cdf_upper_,
+            rng=rng,
+            size=rv_inputs_[0],
+        )
+        truncated_rv_ = icdf(rv_, uniform_)
+        truncated_rv = TruncatedRV(
+            base_rv_op=rv.owner.op,
+            inputs=graph_inputs_,
+            outputs=[uniform_.owner.outputs[0], truncated_rv_],
+            inline=True,
+        )(*graph_inputs)
+        updates = (
+            truncated_rv.owner.inputs[-1],
+            truncated_rv.owner.outputs[0],
+        )
+        return truncated_rv, updates
+    except NotImplementedError:
+        pass
+
+    # Fallback to rejection sampling
+    # TODO: Handle potential broadcast by lower / upper
+    rv_ = rv.owner.op.make_node(rng, *rv_inputs_).default_output()
+
+    def loop_fn(truncated_rv, reject_draws, lower, upper, rng, *rv_inputs):
+        next_rng, new_truncated_rv = rv.owner.op.make_node(rng, *rv_inputs).outputs
+        truncated_rv = at.set_subtensor(
+            truncated_rv[reject_draws],
+            new_truncated_rv[reject_draws],
+        )
+        reject_draws = at.or_((truncated_rv < lower), (truncated_rv > upper))
+
+        return (
+            (truncated_rv, reject_draws),
+            [(rng, next_rng)],
+            until(~at.any(reject_draws)),
+        )
+
+    (truncated_rv_, reject_draws_), updates = scan(
+        loop_fn,
+        outputs_info=[
+            at.empty_like(rv_),
+            at.ones_like(rv_, dtype=bool),
+        ],
+        non_sequences=[lower_, upper_, rng, *rv_inputs_],
+        n_steps=max_n_steps,
+        strict=True,
+    )
+
+    truncated_rv_ = truncated_rv_[-1]
+    convergence_ = ~at.any(reject_draws_[-1])
+    truncated_rv_ = Assert(
```

Comment:
This should probably be its own `CheckAndRaise` subclass that raises something like a `ValueError`.  The expectation is that `Assert`s can generally be ignored in performance-critical code, which is why we have an optional rewrite that removes them.  Since this is a numerical procedure that may or may not converge, it doesn't seem reasonable that one could remove `Assert`s for performance reasons and unknowingly end up using numerically incorrect results.

There are still some places in Aesara where `Assert` is used somewhat incorrectly in this way (e.g. broadcasting, I believe), but the expectation is that the underlying error will necessarily manifest in an `Op.perform` later down the line, so the `Assert` is just a emulating the place and type of error message that NumPy would return, but not avoiding the error altogether.  (Really, the issue here is that those `Assert`s should probably be the exact same exception types that NumPy would raise.)

Reply:
Yeah, I'll make a custom Error for this